dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java

```diff
@@ -0,0 +1,220 @@
+    RexLiteral fetchLiteral = (RexLiteral) fetch;
+    count = ((BigDecimal) fetchLiteral.getValue()).intValue();
+
+    if (offset != null) {
+      RexLiteral offsetLiteral = (RexLiteral) offset;
+      startIndex = ((BigDecimal) offsetLiteral.getValue()).intValue();
+    }
+  }
+
+  @Override public Pipeline buildBeamPipeline(BeamPipelineCreator planCreator) throws Exception {
+    RelNode input = getInput();
+    BeamSQLRelUtils.getBeamRelInput(input).buildBeamPipeline(planCreator);
+
+    String stageName = BeamSQLRelUtils.getStageName(this);
+    PCollection<BeamSQLRow> upstream = planCreator.popUpstream();
+
+    BeamSQLRowComparator comparator = new BeamSQLRowComparator(fieldIndices, orientation);
+    // first find the top (offset + count)
+    PCollection<BeamSQLRow> orderedStream =
+        upstream.apply("extractTopOffset_plus_Fetch",
+            Top.of(startIndex + count, comparator))
+        .apply("extractTopOffset_plus_Fetch__Flatten", Flatten.<BeamSQLRow>iterables());
+
+    if (startIndex > 0) {
+      // strip the leading `offset`
+      orderedStream = orderedStream.apply("stripLeadingOffset",
+          Top.of(count, new NegativeComparator(comparator)))
+          .apply("stripLeadingOffset__Flatten", Flatten.<BeamSQLRow>iterables());
+      // reverse it
+      orderedStream = orderedStream.apply("reverseTailFetch",
+          Top.of(count, comparator))
+          .apply("reverseTailFetch__Flatten", Flatten.<BeamSQLRow>iterables());
+    }
+
+    planCreator.pushUpstream(orderedStream);
+
+    return planCreator.getPipeline();
+  }
+
+  @Override public Sort copy(RelTraitSet traitSet, RelNode newInput, RelCollation newCollation,
+      RexNode offset, RexNode fetch) {
+    return new BeamSortRel(getCluster(), traitSet, newInput, newCollation, offset, fetch);
+  }
+
+  private static class NegativeComparator implements Comparator<BeamSQLRow>, Serializable {
+    private BeamSQLRowComparator delegate;
+    public NegativeComparator(BeamSQLRowComparator delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override public int compare(BeamSQLRow row1, BeamSQLRow row2) {
+      return delegate.compare(row2, row1);
+    }
+  }
+
+  private static class BeamSQLRowComparator implements Comparator<BeamSQLRow>, Serializable {
+    private List<Integer> fieldsIndices;
+    private List<Boolean> orientation;
+    public BeamSQLRowComparator(List<Integer> fieldsIndices, List<Boolean> orientation) {
+      this.fieldsIndices = fieldsIndices;
+      this.orientation = orientation;
+    }
+
+    @Override public int compare(BeamSQLRow row1, BeamSQLRow row2) {
+      for (int i = 0; i < fieldsIndices.size(); i++) {
+        int fieldIndex = fieldsIndices.get(i);
+        int fieldRet = 0;
+        SqlTypeName fieldType = row1.getDataType().getFieldsType().get(fieldIndex);
+        if (row1.isNull(fieldIndex) && row2.isNull(fieldIndex)) {
+          continue;
+        } else if (row1.isNull(fieldIndex) && !row2.isNull(fieldIndex)) {
```

Comment:
Good catch! There is actually dedicated syntax to specify this: `NULL FIRST`, `NULLS LAST`.